virtcontainers/sandbox.go

```diff
@@ -1105,45 +1105,70 @@ func (s *Sandbox) ListRoutes() ([]*grpc.Route, error) {
 
 // startVM starts the VM.
 func (s *Sandbox) startVM() error {
+	var err error
 	span, ctx := s.trace("startVM")
 	defer span.Finish()
 
 	s.Logger().Info("Starting VM")
 
-	// FIXME: This would break cached VMs. We need network hotplug and move
-	// oci hooks and netns handling to cli. See #273.
-	if err := s.network.run(s.networkNS.NetNsPath, func() error {
-		if s.factory != nil {
-			vm, err := s.factory.GetVM(ctx, VMConfig{
+	startSandboxFunc := func() error {
+		return s.hypervisor.startSandbox()
+	}
+	postStartSandboxFunc := func() error {
+		if err := s.agent.waitForAgentReady(s); err != nil {
+			return err
+		}
+		s.Logger().Info("Agent initialization completed")
+		return nil
+	}
+
+	// Use factory if available
+	if s.factory != nil {
+		var vm *VM
+		startSandboxFunc = func() error {
+			if vm, err = s.factory.GetVM(ctx, VMConfig{
 				HypervisorType:   s.config.HypervisorType,
 				HypervisorConfig: s.config.HypervisorConfig,
 				AgentType:        s.config.AgentType,
 				AgentConfig:      s.config.AgentConfig,
-			})
-			if err != nil {
+			}); err != nil {
 				return err
 			}
-			err = vm.assignSandbox(s)
-			if err != nil {
+			if err = vm.assignSandbox(s); err != nil {
```

Comment:
`assignSandbox` should be called **after** `InitializeVM`. The purpose is to assign a fully functional guest to a sandbox.